### ISSUE TITLE
Explore getting rid of EventMachine in favor of going all-in on Celluloid::IO

### ISSUE THREAD

**Issue by dhh** [rails/actioncable#16]
@tarcieri and @jasonayre mentioned on Twitter that we should be able to use [Angelo](https://github.com/kenichi/angelo)/Reel over Celluloid::IO instead of relying on EventMachine. Let's explore that, do the benchmarks, and see if we can't make that happen. Would certainly be nicer to rely on a single concurrency model throughout the framework if that can be shown to work just as well (or better).

**Comment by digitalextremist** [comment]
I've looked at this and seen it before in a `Slack` API gem, and all uses of the [faye-websocket](/faye/faye-websocket-ruby) ( which pulls in (websocket-driver)[/faye/websocket-driver-ruby) ).

The author @jcoglan has worked with us in making adjustments to (Reel)[/celluloid/reel] to suit `faye-*` libraries, and visa versa. Our contributors are pretty involved in `faye-*` gems too.

Over at `Celluloid` we started refactoring out all libraries that depend on `Eventmachine` at all, and for us to be able to provide a release of `Reel` which completes this would not be very hard and is already started. I think @kenichi would agree `Angelo` too would benefit from changes being made to `Reel`, because `Angelo` is a DSL like Rails, Sinatra, etc, except it depends on `Reel` rather than `Rack` plus handlers. `Angelo` has almost been developed from the websocket/sse outward, with a mind for organic concurrency... so it is also troubled by the inability to depend purely on `Celluloid` ...

To remove `Eventmachine` for downstream gems, simplify dependencies and provide better performance, using `Reel` as a dependency makes the most sense. It provides both `Celluloid` and `Celluloid::IO`. It can also provide refactored versions of `faye-*` gems in its next release... or forked and revamped versions ( pending @jcoglan's thoughts on which is best ) ... this will help a lot for many cases, both for websocket server and client implementations, as well as for any DSL.

**Comment by jcoglan** [comment]
I emailed @lifo a while ago suggesting use of websocket-driver instead of faye-websocket. websocket-driver is where all the actual protocol logic lives, in an I/O- and concurrency-agnostic way. faye-websocket just hooks that up to EM and wraps the standard WebSocket API around it. It exists because Faye itself was always based on EM and faye-websocket was extracted from it, and then websocket-driver was extracted from *that*.

**Comment by jcoglan** [comment]
The only thing you might miss from faye-websocket if you drop it are the [monkeypatches](https://github.com/faye/faye-websocket-ruby/tree/master/lib/faye/adapters) it includes for various appservers to make them support raw TCP access. This will hopefully become less necessary over time as more servers support `rack.hijack`.

**Comment by digitalextremist** [comment]
@jcoglan I do see `websocket-driver` is I/O agnostic, with `EventMachine` being a development dependency, probably for the tests.

@dhh it seems `EM.*` calls occur in examples, and for `Redis` support.

@tarcieri, can we use [celluloid-redis](/celluloid/celluloid-redis) to handle that? The WebSocket changes are pretty simple apart from that.

**Comment by jasonayre** [comment]
@digitalextremist I don't see why not, I used celluloid-redis, in combination with reel on ATR (which relied entirely on redis pub/sub for broadcasting from the app server process to the websocket connections)
https://github.com/jasonayre/atr

**Comment by digitalextremist** [comment]
@jasonayre it also powers `DCell` ... I can handle the websocket standardization in celluloid/reel#201, maybe you, @tarcieri, @niamster or @Asmod4n could handle the `celluloid-redis` refactor portion?

**Comment by niamster** [comment]
@digitalextremist I'm quite busy lately, especially going to be busy this weekend. Afterwards I should be able to participate here.

**Comment by digitalextremist** [comment]
@niamster understood. For now, can you just give me your opinion of what you'd prefer to support down the line: I figure the best approach here is to abstract/implement `websocket-driver` similarly to how we do with `Celluloid::ZMQ` as `Reel::WebSocket` ... With that piece in mind I will map out the functionality needed and start an extraction gem and a patch to use that gem on this issue. If you feel confident with `celluloid-redis` since you refactored `DCell` recently, it'd be most helpful if you could start on the `Redis` side when you have time, and I will start on the `WebSocket` side. Sound good?

**Comment by digitalextremist** [comment]
@tarcieri I've spoken with @kenichi about this and he and I will talk more about it. For now here's the gist of what seems like a solid approach to get the most done for everyone at once:

At its heart, `Rails` like `Angelo` and like `Sinatra` is a DSL. Two of those three depend on `Rack` for their `HTTP/S` layer, but suffer in the `WS/S` area because they are not descendants of an evented concurrent I/O suite, whereas `Angelo` is the 'grandchild' of `Celluloid::IO` ... It makes sense for `Reel` ( 'parent' of `Angelo`, 'child' of `Celluloid::IO` ) to provide just that aspect in a modular way, so `Reel` can become to `WS/S` ( and `HTTP/S` if desired ) what `Celluloid::IO` is on a more abstract level.

What we're thinking is to extract the websocket aspect of `Reel` so `Reel` itself depends on it as of `0.6.5`, and so DSL's and frameworks/servers can bolt on the only piece they do not do: evented concurrent I/O.

From there, along with providing patches for `Rails` and `Sinatra` to use `Reel::WebSocket` ( Perhaps `Reel::IO` would be more appropriate with `SSE` *and* `WS/S` and more coming ) we will also match the existing Ruby DSL's out there by implementing in `Angelo` and in another DSL yet to be extracted. This way, in the future, in situations like this thread, pieces can be taken out of the demonstration DSLs, and there will be a lot of tests to reuse and modular framework pieces to go around.

**Comment by tarcieri** [comment]
Seems ok?

**Comment by digitalextremist** [comment]
Right on.

**Comment by kenichi** [comment]
For actioncable, I think the first approach is to try to get celluloid-redis working in place of EM::Hiredis. That would be a good start and provide a solid vantage point for future actioncable work, which may or may not include Reel and/or other parts of the Celluloid world.

I tried to uncomment the tests and make them work last night; alas. I then tried to follow the readme and make a rails app with actioncable, but there are many moving parts, and my rails skills are very rusty.

**Comment by digitalextremist** [comment]
Divide and conquer? We were both typing at the same time. I am comfortable with the SSE&WS/S angle. Posted an issue above where I'll be starting in the next 24-72 hours. If I keep to the SSE&WS/S side versus Redis, I can hit much of this on the way to Reel `0.6.5`

**Comment by ianks** [comment]
Surely https://github.com/ruby-concurrency/concurrent-ruby should be in the discussion here, especially since there have been discussions to use it for other parts of Rails (lock free caches, etc). Concurrent-Ruby provides an actor implementation, as well as many other concurrency abstractions.

On that note, if the decision to use an actor based concurrency model is made (which IMO is worthy of a discussion in and of itself), the implementation of the actor should be configurable for the user.

**Comment by tarcieri** [comment]
@ianks concurrent-ruby isn't really germane to the discussion because it doesn't have evented I/O support. Don't get me wrong, I liked it and used it in [Turbine](https://github.com/tarcieri/turbine), but it has nothing to do with the problems being discussed in this thread.

**Comment by niamster** [comment]
@digitalextremist I can do refactoring of `celluloid-redis`. Let's talk about it on Monday if you don't mind ..

**Comment by tarcieri** [comment]
@niamster there are some issues with the existing `celluloid-redis` API (namely it defines a class named `Redis` inside the `Celluloid` module which is not a subclass or duck type of `::Redis`). It'd be great if you could address those problems too.

**Comment by dre-hh** [comment]
@ianks @tarcieri 
Plz correct me if i'm wrong, but as far as I understand the idea of concurrent-ruby is to implement light-weight actors like in Erlang. So there are multiple actors per native OS thread and separate Scheduler within concurrent-ruby. Those actor can still use synchronous IO libs and they get rescheduled when they block.  This paradigm allows Erlang to spawn millions of those lightweight processes and does not require you to deal with async callback code.

**Comment by tarcieri** [comment]
@dre-hh this particular thread is about replacing EventMachine with Celluloid::IO. If you think there's a blocking I/O strategy that makes sense to replace EventMachine with (in addition to replacing Celluloid with concurrent-ruby, I guess?), perhaps you could make another thread for that.

All that said I think Celluloid::IO is a better fit given what's already in place. Celluloid::IO lets actors multiplex IO operations with actor messages, so it enables a much wider set of use cases than is possible with a purely blocking I/O strategy.

**Comment by dre-hh** [comment]
@tarcieri 
>Celluloid::IO is a better fit given what's already in place
concurrent-ruby actors seem also slight far from being production ready
https://github.com/ruby-concurrency/concurrent-ruby/issues/332

**Comment by Asmod4n** [comment]
Last Time i checked, the actor implementation of concurrent-ruby was using a very tiny thread pool which shouldn't be used for blocking I/O because it would exhaust the thread pool rather quickly.
Or you would have to create your own pool and know in advance how many sockets you would open.

But thats over one year ago and i haven't seen any advance in support for I/O in concurrent-ruby.

**Comment by SamSaffron** [comment]
Keep in mind, the only reason you need EM here is for timer implementation (so sockets eventually clean up if rudely disconnected), this can be done in native ruby quite efficiently  

https://github.com/SamSaffron/message_bus/blob/master/lib/message_bus/timer_thread.rb

**Comment by Asmod4n** [comment]
@SamSaffron Time.new shouldn't be used to calculate timer offsets because the System time doesn't increase monotonically, but ruby doesn't offer a monotonic clock, right?

**Comment by ianks** [comment]
concurrent-ruby has monotonic time https://github.com/ruby-concurrency/concurrent-ruby/blob/4a64e82a392bbdad4c39e9d873fb6863b6f40719/lib/concurrent/utility/monotonic_time.rb

**Comment by digitalextremist** [comment]
@ianks, @SamSaffron the scope is to fully take advantage of what's already in place elsewhere in the project, consolidating concurrency approaches since the existing majority of the code is already evented, and it has evented IO by bringing in a lightweight supporting library... rather than have multiple concurrency tending libraries working in parallel. This gem already uses `Celluloid` for its `Worker` actor... now it'll use it for `Redis` and `WebSocket` management. With the way `Celluloid` works, these other two areas just add actors to the existing actor system already running, using clocks and other functionality already there.

**Comment by tarcieri** [comment]
> the only reason you need EM here is for timer implementation

Well, that and an event-based WebSockets implementation, right?

That said, while we're talking about monotonic time, Celluloid uses the hitimes and timers gems to implement monotonic time and timers respectively.

I'll go back to my original comment:

> If you think there's a blocking I/O strategy that makes sense to replace EventMachine with (in addition to replacing Celluloid with concurrent-ruby, I guess?), perhaps you could make another thread for that.

Seems a lot easier said than done to me...

**Comment by tarcieri** [comment]
> timer implementation (so sockets eventually clean up if rudely disconnected), this can be done in native ruby quite efficiently
>
> https://github.com/SamSaffron/message_bus/blob/master/lib/message_bus/timer_thread.rb

Beyond just this, you need some way of unblocking I/O operations that can last an indefinite amount of time. This is where Ruby traditionally uses something like `timeout.rb`, but `timeout.rb is inherently racy.

To actually solve the races and implement timeouts correctly, you need an evented I/O subsystem.

**Comment by digitalextremist** [comment]
Have a demo/broken/untested/unused initial push here. I haven't run it even once:

* https://github.com/celluloid/reel-io

After having worked with these types of connections for a long time, making an extremely simple API possible seemed like a good opportunity.

The commits I have so far still need to port `EventSource` since that doesn't appear in `WebSocket::Driver` as it stands, and it's mostly a preview of syntax... but not far from usable since the `Reel` implementation of `WebSocket::Driver` is pretty standardized.

Will be back to take 2-4 more stabs at it until it's done. There shouldn't be any surprises since it's basically just a wrapper for `Celluloid::IO` and `WebSocket::Driver` with some life enhancements.

**Comment by jdantonio** [comment]
As others have suggested, this thread probably isn't the best place to discuss [concurrent-ruby](https://github.com/ruby-concurrency/concurrent-ruby). I've created a [discussion thread](https://github.com/ruby-concurrency/concurrent-ruby/issues/379) on our repo where myself and the others on the team will be happy to answer any questions.

Since concurrent-ruby is now a [runtime dependency of Rails](http://weblog.rubyonrails.org/2015/7/17/this-week-in-rails-goodbye-rack-lock-concurrent-ruby-better-ac-parameters-and-more/) (activesupport and activerecord) and has been discussed on this thread, I'd like to take a moment to provide clarification on a few specific comments. I don't want to hijack this thread so I'll ask that any follow-up questions/comments be moved to the [aforementioned thread](https://github.com/ruby-concurrency/concurrent-ruby/issues/379) on the concurrent-ruby repo.

**Clarifications:**

> there have been discussions to use it for other parts of Rails

- @ianks

Correct. As of last week, concurrent-ruby is a runtime dependency of Edge Rails (activesupport and activerecord). We've replaced the Rails countdown latch with our own, are actively working to replace Rails' read-write lock, and are looking into other ways we can help to optimize and enhance Rails. Additionally, the [thread_safe](https://github.com/ruby-concurrency/thread_safe) gem, which has been a runtime dependency of Rails for a long time, is now part of the [ruby-concurrency](https://github.com/ruby-concurrency) family. It will be [deprecated and merged into concurrent-ruby](https://github.com/ruby-concurrency/thread_safe/issues/2), hopefully by the end of the year.

> concurrent-ruby... doesn't have evented I/O support like Celluloid::IO does

- @tarcieri

Correct. We do not have any evented I/O support right now and it is not on our immediate roadmap.

> the idea of concurrent-ruby is to implement light-weight actors like in Erlang

- @dre-hh

concurrent-ruby is intended to be a rich concurrency library that includes both high-level abstractions (inspired by Clojure, Scala, Erlang, and other languages) and low-level tools (mostly inspired by java.util.concurrent). The main inspiration for our actors is [Akka](http://akka.io/). A list of the tools we provide can be found in the [readme](https://github.com/ruby-concurrency/concurrent-ruby/blob/master/README.md#features--documentation).

> concurrent-ruby actors seem also slight far from being production ready

- @dre-hh

Our actors are currently being used in production both directly and also though projects like [Dynflow](https://github.com/Dynflow/dynflow). Our current actor implementation depends on our new future/promise API. This is a new API that is significantly different from the current future/promise API. Our roadmap is to release 1.0 with the original API (within the next month or so) then quickly release 2.0 with the new API. This will allow existing users to upgrade to 2.0 at their leisure, rather than forcing the new API on them at 1.0.

> the actor implementation of concurrent-ruby was using a very tiny thread pool which shouldn't be used for blocking I/O

- @Asmod4n

There has been significant development with our actors and the current implementation works very well with blocking I/O operations. @pitr-ch is the main implementer and will happily answer actor questions.

> Or you would have to create your own pool and know in advance...

- @Asmod4n

All of the high-level abstraction in concurrent-ruby support thread-pool configuration via dependency injection. Additionally, our thread pools are highly configurable (based heavily on the thread pools in java.util.concurrent). This can be very advantageous to library/framework authors using concurrent-ruby, but the default thread pools are configured for the types of blocking I/O operations performed by many Ruby applications.

> concurrent-ruby has monotonic time

- @ianks

Correct. That was a contribution from @tenderlove, with later enhancements from others.

> the scope is to fully take advantage of what's already in place elsewhere in the project

- @digitalextremist

ActionCable lists activesupport as a dependency in its gemspec. concurrent-ruby is a [runtime dependency](https://github.com/rails/rails/blob/master/activesupport/activesupport.gemspec) of activerecord in Edge Rails (this was [announced last week](http://weblog.rubyonrails.org/2015/7/17/this-week-in-rails-goodbye-rack-lock-concurrent-ruby-better-ac-parameters-and-more/) on the Riding Rails blog). Eventually actioncable will have access to concurrent-ruby via that dependency chain. If there are any utilities within concurrent-ruby that may assist this project I encourage you to explore them.

**Comment by Asmod4n** [comment]
@jdantonio thats cool :)

**Comment by Zapotek** [comment]
I came to the ActionCable repo to specifically check for EventMachine use, aside from the overall niceness of relying on a single concurrency model, EM has had issues with Windows support; last time I used (gave up on) EM it didn't install on Windows.

CelluloidIO seems to have some [Windows issues](https://github.com/celluloid/celluloid-io/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+windows) as well.

I know supporting Windows isn't fun, but it's important to a lot of people so please keep it in mind.

**Comment by tarcieri** [comment]
Fixing things Windows-related issues on Celluloid::IO should be generally easier than on EventMachine as the overwhelming majority of the code is written in Ruby. Furthermore, two of the issues are related to DNS, and there are plans to switch to the much more robust RubyDNS resolver which I hope would address those issues.

**Comment by dhh** [comment]
Doesn't seem like any additional progress has been made on this. So closing for now. If anyone wants to do some work exploring this, please open a PR.

**Comment by digitalextremist** [comment]
@dhh will open a PR with existing code once completed and thoroughly tested.
